timemachine/cpp/src/neighborlist.cu

```diff
@@ -99,13 +100,21 @@ std::vector<std::vector<int>> Neighborlist<RealType>::get_nblist_host(
     } else if (h_row_coords == nullptr && NR != 0) {
         throw std::runtime_error("No row coords provided, but NR != 0");
     }
-    double *d_col_coords = gpuErrchkCudaMallocAndCopy(h_column_coords, NC * 3);
 
-    double *d_row_coords = this->compute_full_matrix() ? gpuErrchkCudaMallocAndCopy(h_row_coords, NR * 3) : nullptr;
+    bool store_row = this->compute_full_matrix();
+
+    DeviceBuffer<double> d_col_coords(NC * 3);
+    DeviceBuffer<double> d_box(3 * 3);
+    DeviceBuffer<double> d_row_coords(store_row ? NR * 3 : 0);
```

Comment:
nit: this changes the behavior to always allocate. It probably won't make much difference, but I think we can use `std::unique_ptr<DeviceBuffer<double>>` to recover the old behavior (e.g. like we do in [potential.cu](https://github.com/proteneer/timemachine/blob/776873a013e65f8d825570312d0fa87d3a438522/timemachine/cpp/src/potential.cu#L36-L57))

Reply:
I would've thought that a malloc/free of size 0 should be a no-op. However, there may be an issue with the resulting `data` ptr inside a `DeviceBuffer`. Semantically, `execute_device` code expects the device points to be a `nullptr` to indicate that an operation want to be skipped. The current `allocate` function of the `DeviceBuffer` class does:

```
template <typename T> T *allocate(const std::size_t length) {
    T *buffer;
    gpuErrchk(cudaMalloc(&buffer, length * sizeof(T)));
    return buffer;
}
```

So if `length` is zero, I don't think this returns a `nullptr` (rather it just returns a random pointer that points to a zero-sized chunk). Though I'm not sure what the correct behavior is:

1) return a `nullptr` if `length` is zero.
2) return a non-nullptr

What do you guys think?

Reply:
I would say it should be `nullptr` as that is our convention in the rest of the code for "don't compute anything for this"

Reply:
It is actually already a `nullptr`, looks like cudaMalloc is a noop if the size is 0. Will make it more explicit to be safe

Reply:
technically if `cudaMalloc` sets `buffer` to a `nullptr` then it's not quite a `no-op` since the pointer will have been changed.

Reply:
I find it convenient and succinct to be able to say `DeviceBuffer(0)` rather than messing with a unique pointer. And since we own the code, and are explicitly not allocating I don't see it as a code smell.

Reply:
Thinking about it more, going to follow the unique ptr style and as add a runtime warning to device buffer if you set the size to <= 0